### PR TITLE
[DTensor] Fix backward support for cumprod, cummax, cummin

### LINKED ISSUE #185362: [DTensor] backward fails for cumprod, cummax, cummin
Added pr fixing this

repro
```
import os
import sys

import torch
import torch.distributed as dist
from torch.distributed.tensor import Replicate, Shard, distribute_tensor, init_device_mesh

dist.init_process_group("nccl")
torch.cuda.set_device(int(os.environ["LOCAL_RANK"]))
mesh = init_device_mesh("cuda", (dist.get_world_size(),))

op = sys.argv[1]

if op == "cumprod":
    x_local = torch.rand(12, 8, device="cuda").add_(0.1)
    x_local.select(0, x_local.size(0) // 2).zero_()
    x = distribute_tensor(x_local.requires_grad_(True), mesh, [Shard(0)])
    out = torch.cumprod(x, dim=0)
elif op == "masked_cumprod":
    x = distribute_tensor(
        torch.rand(12, 8, device="cuda").add_(0.1).requires_grad_(True),
        mesh,
        [Shard(0)],
    )
    row = torch.arange(12, device="cuda").unsqueeze(1)
    col = torch.arange(8, device="cuda").unsqueeze(0)
    mask = distribute_tensor((row + col) % 2 == 0, mesh, [Replicate()])
    out = torch.masked.cumprod(x, mask=mask, dim=0)
else:
    x = distribute_tensor(
        torch.randn(12, 8, device="cuda").requires_grad_(True),
        mesh,
        [Shard(0)],
    )
    out = getattr(torch, op)(x, dim=0)[0]

out.sum().backward()
print(f"[rank {dist.get_rank()}] {op} backward done", flush=True)

dist.destroy_process_group()
```

backward fails with a mixed Tensor / DTensor error

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @tianyu-l @XilunWu @SherlockNoMad @ppwwyyxx

### PR DESCRIPTION
**Pull request by arkadip-maitra**
Fixes #185362 

`cumprod_backward` and `cummaxmin_backward` were creating plain tensors inside tensor subclass paths, causing mixed tensor error during backward. this change switches those helper allocations to subclass aware functions and changes some ops to out of place for where `DTensor` need composite compliance. also added tests

### CODE REVIEW

File: aten/src/ATen/native/ReduceOps.cpp [hunk 1]
```diff
@@ -908,13 +916,15 @@ Tensor cummaxmin_backward(const Tensor& grad, const Tensor& input, const Tensor&
...
-  // `scatter_add` if `indices` or `grad` is a Tensor Subclass.
+  // for composite compliance and tensor subclasses (e.g. DTensor), create the
+  // scatter destination via the subclass so subsequent scatter_add stays in the
+  // same tensor subclass.
   if (areAnyTensorSubclassLike({indices, grad})) {
+    auto result = grad.new_zeros_symint(input.sym_sizes());
     return result.scatter_add(dim, indices, grad);
   }
+  auto result = at::zeros_symint(input.sym_sizes(), input.options());
   return result.scatter_add_(dim, indices, grad);
```
Comment (Skylion007) [inline aten/src/ATen/native/ReduceOps.cpp:944]: This will break functorch due to a bug I'm pretty sure regarding scatter_add_ support on ZeroTensors

File: aten/src/ATen/native/ReduceOps.cpp [hunk 2]
```diff
@@ -634,6 +634,14 @@ Tensor cumprod_backward(const Tensor& grad, const Tensor& input, int64_t dim, co
 
   // For Composite Compliance, we always choose the slower but composite compliant path.
   bool are_inputs_tensors_sublcass = areAnyTensorSubclassLike({input, grad, output});
+  auto make_subclass_aware_zeros = [&](c10::SymIntArrayRef sizes) {
+    return are_inputs_tensors_sublcass ? grad.new_zeros_symint(sizes)
+                                       : at::zeros_symint(sizes, grad.options());
+  };
+  auto make_subclass_aware_ones = [&](c10::SymIntArrayRef sizes) {
+    return are_inputs_tensors_sublcass ? grad.new_ones_symint(sizes)
+                                       : at::ones({1}, grad.options()).expand_symint(sizes);
```
[the remark this answers is not in the document]
Comment (arkadip-maitra) [reply aten/src/ATen/native/ReduceOps.cpp]: added create_graph=True and higher order tests

File: aten/src/ATen/native/ReduceOps.cpp (hunk 1 above)
Comment (arkadip-maitra) [reply aten/src/ATen/native/ReduceOps.cpp:944]: i checked with torch func and vmap with cummax/cummin and it did not break

File: aten/src/ATen/native/ReduceOps.cpp [hunk 3]
```diff
@@ -752,14 +756,36 @@ Tensor cumprod_backward(const Tensor& grad, const Tensor& input, int64_t dim, co
...
+              at::cumprod(input_conj.slice(dim, 0, k), dim).slice(dim, k - 1, k);
+          grad_slice = grad.select(dim, k) * prods_until_k.squeeze(dim);
+          if (k != dim_size - 1) {
+            prods_from_k_plus_1 = at::cumprod(input_conj.slice(dim, k + 1), dim);
+            const Tensor omitted_products_tail =
+                prods_until_k.expand_as(prods_from_k_plus_1) * prods_from_k_plus_1;
+            grad_slice = grad_slice + at::sum(grad.slice(dim, k + 1) * omitted_products_tail, dim);
+          }
+        }
+        grad_inputs.push_back(grad_slice);
```
Comment (Skylion007) [inline aten/src/ATen/native/ReduceOps.cpp]: Should be std::move here

**Comment by Skylion007** [comment]
@pytorch rebase

File: aten/src/ATen/native/ReduceOps.cpp [hunk 4]
```diff
@@ -644,6 +644,10 @@ Tensor cumprod_backward(const Tensor& grad, const Tensor& input, int64_t dim, co
 
   // For Composite Compliance, we always choose the slower but composite compliant path.
   bool are_inputs_tensors_sublcass = areAnyTensorSubclassLike({input, grad, output});
```
Comment (Skylion007) [inline aten/src/ATen/native/ReduceOps.cpp]: This Boolean has a typo

File: aten/src/ATen/native/ReduceOps.cpp (hunk 3 above)
Comment (Skylion007) [reply aten/src/ATen/native/ReduceOps.cpp]:
```suggestion
        grad_inputs.push_back(std::move(grad_slice));

File: aten/src/ATen/native/ReduceOps.cpp [hunk 5]
```diff
@@ -708,8 +712,8 @@ Tensor cumprod_backward(const Tensor& grad, const Tensor& input, int64_t dim, co
     // which is true in the range [z1, z2)
     const auto first_zero_index = std::get<1>(mask.max(dim, /*keepdim*/ true));
     const auto first_zero_mask = at::zeros_like(mask)
```
Comment (Skylion007) [inline aten/src/ATen/native/ReduceOps.cpp]: I really hate this paradigm, I wish there was a way to do inplace IF not DTensor, sigh.

File: aten/src/ATen/native/ReduceOps.cpp (hunk 1 above)
Comment (Skylion007) [reply aten/src/ATen/native/ReduceOps.cpp:944]: Ah< I see because it only breaks on the subclass path

File: aten/src/ATen/native/ReduceOps.cpp [hunk 6]
```diff
@@ -912,13 +934,15 @@ Tensor cummaxmin_backward(const Tensor& grad, const Tensor& input, const Tensor&
...
-  auto result = at::zeros_symint(input.sym_sizes(), input.options());
 
-  // for composite compliance, use out-of-place variant of
-  // `scatter_add` if `indices` or `grad` is a Tensor Subclass.
+  // for composite compliance and tensor subclasses (e.g. DTensor), create the
+  // scatter destination via the subclass so subsequent scatter_add stays in the
+  // same tensor subclass.
   if (areAnyTensorSubclassLike({indices, grad})) {
+    auto result = grad.new_zeros_symint(input.sym_sizes());
     return result.scatter_add(dim, indices, grad);
```
Comment (Skylion007) [inline aten/src/ATen/native/ReduceOps.cpp:941]: Opened a PR fixing scatter_add_

File: aten/src/ATen/native/ReduceOps.cpp (hunk 4 above)
Comment (arkadip-maitra) [reply aten/src/ATen/native/ReduceOps.cpp]: fixed this, thanks for the comment

File: aten/src/ATen/native/ReduceOps.cpp (hunk 5 above)
Comment (Skylion007) [reply aten/src/ATen/native/ReduceOps.cpp]: Why does this need to happen? Is DTensor missing one of these workflows from autogen?

File: aten/src/ATen/native/ReduceOps.cpp (hunk 6 above)
Comment (Skylion007) [reply aten/src/ATen/native/ReduceOps.cpp:941]: Oh it's probably not related to DTensor at all, sigh...

File: aten/src/ATen/native/ReduceOps.cpp [hunk 7]
```diff
@@ -708,8 +712,8 @@ Tensor cumprod_backward(const Tensor& grad, const Tensor& input, int64_t dim, co
     // which is true in the range [z1, z2)
     const auto first_zero_index = std::get<1>(mask.max(dim, /*keepdim*/ true));
     const auto first_zero_mask = at::zeros_like(mask)
-                                  .scatter_(dim, first_zero_index, /*src*/ 1)
-                                  .logical_and_(mask);
+                                  .scatter(dim, first_zero_index, /*value*/ true)
```
Comment (Skylion007) [inline aten/src/ATen/native/ReduceOps.cpp]: Why are making these out of place necessary?

**Review by Skylion007 — APPROVED** [review]
I just have one last nite about why scatter_ and logical_and_ needed to be out of place?

File: aten/src/ATen/native/ReduceOps.cpp [hunk 8]
```diff
@@ -708,8 +712,8 @@ Tensor cumprod_backward(const Tensor& grad, const Tensor& input, int64_t dim, co
...
+                                  .logical_and(mask);
```
Comment (Skylion007) [inline aten/src/ATen/native/ReduceOps.cpp]:
```suggestion
    const auto first_zero_mask = mask.logical_and(is_zero);
```
This is equivalent, isn't it?

File: aten/src/ATen/native/ReduceOps.cpp (hunk 7 above)
Comment (arkadip-maitra) [reply aten/src/ATen/native/ReduceOps.cpp]: changed it thanks for the suggestion

File: aten/src/ATen/native/ReduceOps.cpp (hunk 8 above)
Comment (arkadip-maitra) [reply aten/src/ATen/native/ReduceOps.cpp]: yes thanks for the suggestion
